File: qqbot-client/src/main/java/com/mieai/qqbot/client/FileMediaAssetStore.java
package com.mieai.qqbot.client;

import com.mieai.qqbot.domain.bot.BotId;
import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.time.Clock;
import java.time.Instant;
import java.util.Locale;
import java.util.Objects;
import java.util.Optional;
import java.util.Properties;
import java.util.UUID;

/** Atomic file-backed media staging with a hard byte limit while reading. */
public final class FileMediaAssetStore implements MediaAssetStore {
    private final Path root;
    private final Clock clock;

    public FileMediaAssetStore(Path root) { this(root, Clock.systemUTC()); }

    public FileMediaAssetStore(Path root, Clock clock) {
        this.root = Objects.requireNonNull(root, "root must not be null").toAbsolutePath().normalize();
        this.clock = Objects.requireNonNull(clock, "clock must not be null");
    }

    @Override
    public MediaAsset stage(BotId botId, QqMediaKind kind, String fileName, String contentType,
            InputStream input, long maxBytes) {
        Objects.requireNonNull(botId, "botId must not be null");
        Objects.requireNonNull(kind, "kind must not be null");
        Objects.requireNonNull(input, "input must not be null");
        if (maxBytes < 1L) throw new IllegalArgumentException("maxBytes must be positive");
        validateName(fileName);
        validateContentType(contentType);
        validateKind(kind, contentType);
        UUID id = UUID.randomUUID();
        Path directory = root.resolve(botId.toString()).normalize();
        Path target = directory.resolve(id + ".bin").normalize();
        if (!target.startsWith(directory)) throw new IllegalArgumentException("invalid media path");
        Path temporary = directory.resolve(id + ".tmp").normalize();
        Path metadata = directory.resolve(id + ".meta").normalize();
        Path temporaryMetadata = directory.resolve(id + ".meta.tmp").normalize();
        long size = 0L;
        String normalizedName = fileName.strip();
        String normalizedContentType = contentType.strip().toLowerCase(Locale.ROOT);
        Instant createdAt = clock.instant();
        MediaAsset result;
        try {
            Files.createDirectories(directory);
            try (InputStream source = input; var output = Files.newOutputStream(temporary,
                    StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)) {
                byte[] buffer = new byte[8192];
                int read;
                while ((read = source.read(buffer)) != -1) {
                    size += read;
                    if (size > maxBytes) throw new MediaSizeExceededException(maxBytes);
                    output.write(buffer, 0, read);
                }
            }
            result = new MediaAsset(id, botId, kind, normalizedName, normalizedContentType, size, createdAt);
            writeMetadata(temporaryMetadata, result);
            move(temporary, target);
            move(temporaryMetadata, metadata);
            return result;
        } catch (IOException exception) {
            deleteQuietly(temporary);
            deleteQuietly(temporaryMetadata);
            deleteQuietly(target);
            deleteQuietly(metadata);
            throw new IllegalStateException("Unable to stage media asset", exception);
        } catch (RuntimeException exception) {
            deleteQuietly(temporary);
            deleteQuietly(temporaryMetadata);
            deleteQuietly(target);
            deleteQuietly(metadata);
            throw exception;
        }
    }

    @Override
    public Optional<MediaAsset> find(BotId botId, UUID id) {
        Objects.requireNonNull(botId, "botId must not be null");
        Objects.requireNonNull(id, "id must not be null");
        Path path = path(botId, id);
        try {
            if (!Files.isRegularFile(path)) return Optional.empty();
            long size = Files.size(path);
            Path metadata = metadataPath(botId, id);
            if (!Files.isRegularFile(metadata)) {
                return Optional.of(new MediaAsset(id, botId, QqMediaKind.FILE, id + ".bin",
                        "application/octet-stream", size,
                        Instant.ofEpochMilli(Files.getLastModifiedTime(path).toMillis())));
            }
            Properties properties = new Properties();
            try (InputStream input = Files.newInputStream(metadata, StandardOpenOption.READ)) {
                properties.load(input);
            }
            QqMediaKind kind = QqMediaKind.valueOf(properties.getProperty("kind"));
            String fileName = properties.getProperty("fileName");
            String contentType = properties.getProperty("contentType");
            long recordedSize = Long.parseLong(properties.getProperty("sizeBytes"));
            Instant createdAt = Instant.parse(properties.getProperty("createdAt"));
            if (recordedSize != size) return Optional.empty();
            return Optional.of(new MediaAsset(id, botId, kind, fileName, contentType, size, createdAt));
        } catch (IOException exception) {
            throw new IllegalStateException("Unable to inspect media asset", exception);
        } catch (RuntimeException exception) {
            throw new IllegalStateException("Unable to inspect media asset metadata", exception);
        }
    }

    @Override
    public InputStream open(MediaAsset asset) {
        Objects.requireNonNull(asset, "asset must not be null");
        try { return Files.newInputStream(path(asset.botId(), asset.id()), StandardOpenOption.READ); }
        catch (IOException exception) { throw new IllegalStateException("Unable to open media asset", exception); }
    }

    @Override
    public void delete(MediaAsset asset) {
        Objects.requireNonNull(asset, "asset must not be null");
        try {
            Files.deleteIfExists(path(asset.botId(), asset.id()));
            Files.deleteIfExists(metadataPath(asset.botId(), asset.id()));
        }
        catch (IOException exception) { throw new IllegalStateException("Unable to delete media asset", exception); }
    }

    private Path path(BotId botId, UUID id) {
        Path directory = root.resolve(botId.toString()).normalize();
        Path path = directory.resolve(id + ".bin").normalize();
        if (!path.startsWith(directory)) throw new IllegalArgumentException("invalid media path");
        return path;
    }

    private Path metadataPath(BotId botId, UUID id) {
        Path directory = root.resolve(botId.toString()).normalize();
        Path path = directory.resolve(id + ".meta").normalize();
        if (!path.startsWith(directory)) throw new IllegalArgumentException("invalid media path");
        return path;
    }

    private static void writeMetadata(Path path, MediaAsset asset) throws IOException {
        Properties properties = new Properties();
        properties.setProperty("kind", asset.kind().name());
        properties.setProperty("fileName", asset.fileName());
        properties.setProperty("contentType", asset.contentType());
        properties.setProperty("sizeBytes", Long.toString(asset.sizeBytes()));
        properties.setProperty("createdAt", asset.createdAt().toString());
        try (var output = Files.newOutputStream(path, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)) {
            properties.store(output, "QQBot media asset metadata");
        }
    }

    private static void move(Path source, Path target) throws IOException {
        try {
            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE);
        } catch (java.nio.file.AtomicMoveNotSupportedException exception) {
            Files.move(source, target);
        }
    }

    private static void deleteQuietly(Path path) {
        try { Files.deleteIfExists(path); } catch (IOException ignored) { }
    }

    private static void validateName(String value) {
        if (value == null || value.isBlank() || value.length() > 255 || value.contains("..")
                || value.indexOf('/') >= 0 || value.indexOf('\\') >= 0
                || value.codePoints().anyMatch(Character::isISOControl)) throw new IllegalArgumentException("fileName is invalid");
    }

    private static void validateContentType(String value) {
        if (value == null || value.isBlank() || value.length() > 128 || value.codePoints().anyMatch(Character::isWhitespace)) {
            throw new IllegalArgumentException("contentType is invalid");
        }
    }

    private static void validateKind(QqMediaKind kind, String contentType) {
        String normalized = contentType.toLowerCase(Locale.ROOT);
        boolean valid = switch (kind) {
            case IMAGE -> normalized.startsWith("image/");
            case VIDEO -> normalized.startsWith("video/");
            case AUDIO -> normalized.startsWith("audio/");
            case FILE -> normalized.startsWith("application/") || normalized.startsWith("text/")
                    || normalized.equals("application/octet-stream");
        };
        if (!valid) throw new IllegalArgumentException("contentType does not match media kind");
    }

    public static final class MediaSizeExceededException extends IllegalArgumentException {
        private final long maxBytes;
        public MediaSizeExceededException(long maxBytes) { super("media upload exceeds " + maxBytes + " bytes"); this.maxBytes = maxBytes; }
        public long maxBytes() { return maxBytes; }
    }
}

File: qqbot-client/src/main/kotlin/com/mieai/qqbot/client/FileMediaAssetStore.kt
package com.mieai.qqbot.client

import com.mieai.qqbot.domain.bot.BotId
import java.io.IOException
import java.io.InputStream
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.StandardOpenOption
import java.time.Clock
import java.time.Instant
import java.util.Locale
import java.util.Properties
import java.util.UUID

/** Atomic file-backed media staging with a hard byte limit while reading. */
class FileMediaAssetStore(
    root: Path,
    private val clock: Clock = Clock.systemUTC(),
) : MediaAssetStore {
    private val root: Path = root.toAbsolutePath().normalize()

    override fun stage(
        botId: BotId,
        kind: QqMediaKind,
        fileName: String,
        contentType: String,
        input: InputStream,
        maxBytes: Long,
    ): MediaAsset {
        require(maxBytes >= 1L) { "maxBytes must be positive" }
        validateName(fileName)
        validateContentType(contentType)
        validateKind(kind, contentType)

        val id = UUID.randomUUID()
        val directory = root.resolve(botId.toString()).normalize()
        val target = directory.resolve("$id.bin").normalize()
        require(target.startsWith(directory)) { "invalid media path" }
        val temporary = directory.resolve("$id.tmp").normalize()
        val metadata = directory.resolve("$id.meta").normalize()
        val temporaryMetadata = directory.resolve("$id.meta.tmp").normalize()
        var size = 0L
        val normalizedName = fileName.trim()
        val normalizedContentType = contentType.trim().lowercase(Locale.ROOT)
        val createdAt = clock.instant()

        try {
            Files.createDirectories(directory)
            input.use { source ->
                Files.newOutputStream(
                    temporary,
                    StandardOpenOption.CREATE_NEW,
                    StandardOpenOption.WRITE,
                ).use { output ->
                    val buffer = ByteArray(8192)
                    while (true) {
                        val read = source.read(buffer)
                        if (read == -1) break
                        size += read
                        if (size > maxBytes) throw MediaSizeExceededException(maxBytes)
                        output.write(buffer, 0, read)
                    }
                }
            }
            val result = MediaAsset(
                id,
                botId,
                kind,
                normalizedName,
                normalizedContentType,
                size,
                createdAt,
            )
            writeMetadata(temporaryMetadata, result)
            move(temporary, target)
            move(temporaryMetadata, metadata)
            return result
        } catch (exception: IOException) {
            deleteQuietly(temporary)
            deleteQuietly(temporaryMetadata)
            deleteQuietly(target)
            deleteQuietly(metadata)
            throw IllegalStateException("Unable to stage media asset", exception)
        } catch (exception: RuntimeException) {
            deleteQuietly(temporary)
            deleteQuietly(temporaryMetadata)
            deleteQuietly(target)
            deleteQuietly(metadata)
            throw exception
        }
    }

    override fun find(botId: BotId, id: UUID): MediaAsset? {
        val path = path(botId, id)
        try {
            if (!Files.isRegularFile(path)) return null
            val size = Files.size(path)
            val metadata = metadataPath(botId, id)
            if (!Files.isRegularFile(metadata)) {
                return MediaAsset(
                    id,
                    botId,
                    QqMediaKind.FILE,
                    "$id.bin",
                    "application/octet-stream",
                    size,
                    Instant.ofEpochMilli(Files.getLastModifiedTime(path).toMillis()),
                )
            }
            val properties = Properties()
            Files.newInputStream(metadata, StandardOpenOption.READ).use(properties::load)
            val kind = QqMediaKind.valueOf(properties.getProperty("kind"))
            val fileName = properties.getProperty("fileName")
            val contentType = properties.getProperty("contentType")
            val recordedSize = properties.getProperty("sizeBytes").toLong()
            val createdAt = Instant.parse(properties.getProperty("createdAt"))
            if (recordedSize != size) return null
            return MediaAsset(id, botId, kind, fileName, contentType, size, createdAt)
        } catch (exception: IOException) {
            throw IllegalStateException("Unable to inspect media asset", exception)
        } catch (exception: RuntimeException) {
            throw IllegalStateException("Unable to inspect media asset metadata", exception)
        }
    }

    override fun open(asset: MediaAsset): InputStream {
        return try {
            Files.newInputStream(path(asset.botId, asset.id), StandardOpenOption.READ)
        } catch (exception: IOException) {
            throw IllegalStateException("Unable to open media asset", exception)
        }
    }

    override fun delete(asset: MediaAsset) {
        try {
            Files.deleteIfExists(path(asset.botId, asset.id))
            Files.deleteIfExists(metadataPath(asset.botId, asset.id))
        } catch (exception: IOException) {
            throw IllegalStateException("Unable to delete media asset", exception)
        }
    }

    private fun path(botId: BotId, id: UUID): Path {
        val directory = root.resolve(botId.toString()).normalize()
        val path = directory.resolve("$id.bin").normalize()
        require(path.startsWith(directory)) { "invalid media path" }
        return path
    }

    private fun metadataPath(botId: BotId, id: UUID): Path {
        val directory = root.resolve(botId.toString()).normalize()
        val path = directory.resolve("$id.meta").normalize()
        require(path.startsWith(directory)) { "invalid media path" }
        return path
    }

    companion object {
        private fun writeMetadata(path: Path, asset: MediaAsset) {
            val properties = Properties()
            properties.setProperty("kind", asset.kind.name)
            properties.setProperty("fileName", asset.fileName)
            properties.setProperty("contentType", asset.contentType)
            properties.setProperty("sizeBytes", asset.sizeBytes.toString())
            properties.setProperty("createdAt", asset.createdAt.toString())
            Files.newOutputStream(path, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE).use {
                properties.store(it, "QQBot media asset metadata")
            }
        }

        private fun move(source: Path, target: Path) {
            try {
                Files.move(source, target, StandardCopyOption.ATOMIC_MOVE)
            } catch (_: AtomicMoveNotSupportedException) {
                Files.move(source, target)
            }
        }

        private fun deleteQuietly(path: Path) {
            try {
                Files.deleteIfExists(path)
            } catch (_: IOException) {
                // Cleanup is best effort; the original failure is retained.
            }
        }

        private fun validateName(value: String) {
            require(!value.isBlank() && value.length <= 255 && ".." !in value &&
                '/' !in value && '\\' !in value &&
                value.codePoints().noneMatch(Character::isISOControl)) {
                "fileName is invalid"
            }
        }

        private fun validateContentType(value: String) {
            require(!value.isBlank() && value.length <= 128 &&
                value.codePoints().noneMatch(Character::isWhitespace)) {
                "contentType is invalid"
            }
        }

        private fun validateKind(kind: QqMediaKind, contentType: String) {
            val normalized = contentType.lowercase(Locale.ROOT)
            val valid = when (kind) {
                QqMediaKind.IMAGE -> normalized.startsWith("image/")
                QqMediaKind.VIDEO -> normalized.startsWith("video/")
                QqMediaKind.AUDIO -> normalized.startsWith("audio/")
                QqMediaKind.FILE -> normalized.startsWith("application/") ||
                    normalized.startsWith("text/") || normalized == "application/octet-stream"
            }
            require(valid) { "contentType does not match media kind" }
        }
    }

    class MediaSizeExceededException(private val maxBytesValue: Long) :
        IllegalArgumentException("media upload exceeds $maxBytesValue bytes") {
        fun maxBytes(): Long = maxBytesValue
    }
}
